internal/benchmarks/trace/main_test.go

```diff
@@ -0,0 +1,126 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"net"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"go.opencensus.io/internal/benchmarks/proto"
+	"go.opencensus.io/plugin/ocgrpc"
+)
+
+func TestMain(m *testing.M) {
+	plainServerAddr = registerAndStartServer(plainServer)
+	noTraceNoStatsAddr = registerAndStartServer(noTraceNoStatsServer)
+	noTraceYesStatsAddr = registerAndStartServer(noTraceYesStatsServer)
+	yesTraceNoStatsAddr = registerAndStartServer(yesTraceNoStatsServer)
+	yesTraceYesStatsAddr = registerAndStartServer(yesTraceYesStatsServer)
+
+	defer noTraceNoStatsServer.Stop()
+	defer noTraceYesStatsServer.Stop()
+	defer yesTraceNoStatsServer.Stop()
+	defer yesTraceYesStatsServer.Stop()
+
+	os.Exit(m.Run())
+}
+
+type server int
+
+func (s *server) Ping(ctx context.Context, p *proto.Payload) (*proto.Payload, error) {
+	return &proto.Payload{"Pong"}, nil
+}
+
+func registerAndStartServer(srv *grpc.Server) string {
+	ln, err := randPortListener()
+	if err != nil {
+		log.Fatalf("randPortListener err: %v", err)
+	}
+	proto.RegisterPingServer(srv, new(server))
+	go srv.Serve(ln)
+	return ln.Addr().String()
+}
+
+var (
+	plainServerAddr, noTraceNoStatsAddr, noTraceYesStatsAddr, yesTraceNoStatsAddr, yesTraceYesStatsAddr string
+)
+
+func randPortListener() (ln net.Listener, err error) {
+	for i := 0; i < 1e3; i++ {
+		p := 4000 + int(rand.Float64()*(65536-4000))
+		ln, err = net.Listen("tcp", fmt.Sprintf(":%d", p))
+		if err == nil {
+			return ln, nil
+		}
+	}
+	return
+}
+
+var (
+	noTraceNoStatsClient   = grpc.WithStatsHandler(&ocgrpc.ClientHandler{NoTrace: true, NoStats: true})
+	noTraceYesStatsClient  = grpc.WithStatsHandler(&ocgrpc.ClientHandler{NoTrace: true, NoStats: false})
+	yesTraceNoStatsClient  = grpc.WithStatsHandler(&ocgrpc.ClientHandler{NoTrace: false, NoStats: true})
+	yesTraceYesStatsClient = grpc.WithStatsHandler(&ocgrpc.ClientHandler{NoTrace: false, NoStats: false})
+
+	plainServer            = grpc.NewServer()
+	noTraceNoStatsServer   = grpc.NewServer(grpc.StatsHandler(&ocgrpc.ServerHandler{NoTrace: true, NoStats: true}))
+	noTraceYesStatsServer  = grpc.NewServer(grpc.StatsHandler(&ocgrpc.ServerHandler{NoTrace: true, NoStats: false}))
+	yesTraceNoStatsServer  = grpc.NewServer(grpc.StatsHandler(&ocgrpc.ServerHandler{NoTrace: false, NoStats: true}))
+	yesTraceYesStatsServer = grpc.NewServer(grpc.StatsHandler(&ocgrpc.ServerHandler{NoTrace: false, NoStats: false}))
+)
+
+func runWithConn(b *testing.B, conn *grpc.ClientConn, qps int) {
+	client := proto.NewPingClient(conn)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		qpsIt(qps, func() {
+			pong, err := client.Ping(context.Background(), &proto.Payload{"Ping"})
+			if err != nil {
+				b.Fatalf("Pong #%d err: %v", i, err)
```

Comment:
I don't think you're allowed to call Fatal from another Goroutine (made this mistake myself)